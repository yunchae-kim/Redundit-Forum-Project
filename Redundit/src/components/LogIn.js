import { useState, useEffect } from "react";
import { useHistory } from "react-router-dom";
import { Button } from "@blueprintjs/core";
import "../style/components/LogInSignUp.css";
import httpClient from "../utility/auth.helper";

const LogIn = () => {
  const history = useHistory();
  const [name, setName] = useState("");
  const [password, setPassword] = useState("");
  const [disabled, setDisabled] = useState(true);
  const [lockoutPolicy, setLocked] = useState(false);
  const [lockoutMessage, setLockoutMessage] = useState(0);

  function handleNameChange(event) {
    setName(event.target.value);
  }

  function handlePasswordChange(event) {
    setPassword(event.target.value);
  }

  async function submitLogin() {
    try {
      const lockout = await httpClient.getLockout(name);

      if (Object.keys(lockout).length !== 0) {
        const { locked, timestamp } = lockout.data;
        if (locked) {
          const LOCKOUT_TIME = 300000; // 5 minutes in milliseconds
          const lockedDate = Date.parse(timestamp) - 18000000; // local time difference
          if (Date.now() >= lockedDate + LOCKOUT_TIME) {
            await httpClient.resetLockout(name);
          } else {
            setLocked(true);
            return;
          }
        }
      }
      const response = await httpClient.login(name, password);
      if (response) {
        if (Object.keys(lockout).length !== 0) {
          await httpClient.resetLockout(name);
        }
        history.push("/");
      }
    } catch (err) {
      const lockout = await httpClient.getLockout(name);
      if (Object.keys(lockout).length === 0) {
        await httpClient.addLockout(name);
        setLockoutMessage(1);
      } else {
        const { attempts } = lockout.data;
        if (attempts >= 5) {
          await httpClient.setLockout(name);
        } else {
          await httpClient.updateLockout(name);
          setLockoutMessage(attempts);
        }
      }
    }
  }

  async function handlePress(event) {
    if (event.key === "Enter" && !disabled) {
      submitLogin();
    }
  }

  useEffect(() => {
    if (name && password) {
      setLocked(false);
      setLockoutMessage(0);
      setDisabled(false);
    } else {
      setDisabled(true);
    }
  }, [name, password]);

  return (
    <div className="LI-container">
      <form onSubmit={submitLogin} className="LI-form">
        <div className="LI-title"> Log-In </div>
        <label htmlFor="userEmail" className="LI-entry">
          Username:
          <input
            className="LI-input"
            required
            type="text"
            name="username"
            placeholder="ex. user12345"
            onChange={handleNameChange}
            onKeyPress={handlePress}
          />
        </label>
        <label htmlFor="userPassword" className="LI-entry">
          Password:
          <input
            className="LI-input"
            required
            type="password"
            name="userPassword"
            placeholder="ex. mypassword"
            onChange={handlePasswordChange}
            onKeyPress={handlePress}
          />
        </label>
        <Button className="LI-submitBtn" text="Log In" onClick={() => submitLogin()} disabled={disabled} />
        {lockoutPolicy
          ? <p className="errorMessage">Reached the maximum number of failed attempts!</p>
          : <></>}
        {lockoutMessage > 0
          ? (
            <p className="errorMessage">
              {5 - lockoutMessage}
              &nbsp;attempt/s left before lockout!
            </p>
          )
          : <></> }
        <span className="LI-signUp">
          New? Sign Up
          {" "}
          <a href="signup">HERE</a>
        </span>
      </form>
    </div>
  );
};

export default LogIn;
