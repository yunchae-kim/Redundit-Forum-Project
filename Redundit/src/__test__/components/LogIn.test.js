import {
  waitFor,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import LogIn from "../../components/LogIn";

test("test should LogIn match snapshot", async () => {
  const tree = renderer.create(
    <LogIn />,
  ).toJSON();
  await waitFor(async () => {
    expect(tree).toMatchSnapshot();
  });
});
