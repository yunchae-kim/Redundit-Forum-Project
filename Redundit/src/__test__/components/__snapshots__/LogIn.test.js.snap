// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test should LogIn match snapshot 1`] = `
<div
  className="LI-container"
>
  <form
    className="LI-form"
    onSubmit={[Function]}
  >
    <div
      className="LI-title"
    >
       Log-In 
    </div>
    <label
      className="LI-entry"
      htmlFor="userEmail"
    >
      Username:
      <input
        className="LI-input"
        name="username"
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="ex. user12345"
        required={true}
        type="text"
      />
    </label>
    <label
      className="LI-entry"
      htmlFor="userPassword"
    >
      Password:
      <input
        className="LI-input"
        name="userPassword"
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="ex. mypassword"
        required={true}
        type="password"
      />
    </label>
    <button
      className="bp3-button bp3-disabled LI-submitBtn"
      disabled={true}
      onBlur={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="bp3-button-text"
      >
        Log In
      </span>
    </button>
    <span
      className="LI-signUp"
    >
      New? Sign Up
       
      <a
        href="signup"
      >
        HERE
      </a>
    </span>
  </form>
</div>
`;
